### PR TITLE
New Resource: `azurerm_data_factory_linked_service_synapse`

### LINKED ISSUE #9919: Support for Azure Data Factory Linked Service to Synapse resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

we should add a resource for Azure Data Factory Linked Service to Synapse.

### New or Affected Resource(s)

* azurerm_data_factory_linked_service_synapse

### Potential Terraform Configuration

Using a connection string with password:
```hcl
resource "azurerm_data_factory_linked_service_synapse" "example" {
  name = "adf-link-synapse"
  resource_group_name = azurerm_resource_group.example.name
  data_factory_name = azurerm_data_factory.example.name

  connection_string = "data source=serverhostname;initial catalog=master;user id=testUser;Password=...;Timeout=30;"
}
```

or using a key vault linked service to store the password:

```hcl
resource "azurerm_data_factory_linked_service_synapse" "example" {
  name                = "adf-link-synapse"
  resource_group_name = azurerm_resource_group.example.name
  data_factory_name   = azurerm_data_factory.example.name

  connection_string = "data source=serverhostname;initial catalog=master;user id=testUser;Timeout=30;"
  key_vault_password {
    linked_service_name  = "adf-link-key-vault"
    secret_name = "synapse-connection-string"
  }
}
```

### References

* https://docs.microsoft.com/en-us/azure/data-factory/connector-azure-sql-data-warehouse#linked-service-properties

### PR DESCRIPTION
**Pull request by allantargino**
fix #9919 

This PR introduces the new resource `azurerm_data_factory_linked_service_synapse`.

You can use a connection string with password:
```hcl
resource "azurerm_data_factory_linked_service_synapse" "example" {
  name = "adf-link-synapse"
  resource_group_name = azurerm_resource_group.example.name
  data_factory_name = azurerm_data_factory.example.name

  connection_string = "data source=serverhostname;initial catalog=master;user id=testUser;Password=...;Timeout=30;"
}
```

or use a key vault linked service to store the password:

```hcl
resource "azurerm_data_factory_linked_service_synapse" "example" {
  name                = "adf-link-synapse"
  resource_group_name = azurerm_resource_group.example.name
  data_factory_name   = azurerm_data_factory.example.name

  connection_string = "data source=serverhostname;initial catalog=master;user id=testUser;Timeout=30;"
  key_vault_password_reference {
    linked_service_name  = "adf-link-key-vault"
    password_secret_name = "synapse-connection-string"
  }
}
```

### CODE REVIEW

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+			"resource_group_name": azure.SchemaResourceGroupNameDiffSuppress(),
+
+			"connection_string": {
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: azureRmDataFactoryLinkedServiceConnectionStringDiff,
+				ValidateFunc:     validation.StringIsNotEmpty,
+			},
+
+			"key_vault_password_reference": {
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:63]: I think we can drop reference from the property name, it doesn't add much

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"linked_service_name": {
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:69]: and could we change this to
```suggestion
						"key_vault_name": {
```

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+
+						"password_secret_name": {
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:75]: and this to
```suggestion
						"secret_name": {
```

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"parameters": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:101]: could we valudate this with at least no empty strings?

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+			},
+
+			"annotations": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:109]: and this

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+			},
+
+			"additional_properties": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:117]: and this

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+		Password:         expandAzureKeyVaultPasswordReference(passwordReference),
+	}
+
+	description := d.Get("description").(string)
+
+	sqlDWLinkedService := &datafactory.AzureSQLDWLinkedService{
+		Description:                           &description,
+		AzureSQLDWLinkedServiceTypeProperties: sqlDWProperties,
+		Type:                                  datafactory.TypeAzureSQLDW,
+	}
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:158]: could we nest these declarations? 
```suggestion
	passwordReference := d.Get("key_vault_password_reference").([]interface{})
	
	sqlDWLinkedService := &datafactory.AzureSQLDWLinkedService{
		Description:                           utils.String(d.Get("description").(string)),
		AzureSQLDWLinkedServiceTypeProperties: &datafactory.AzureSQLDWLinkedServiceTypeProperties{
			ConnectionString: d.Get("connection_string").(string),
			Password:         expandAzureKeyVaultPasswordReference(passwordReference),
		},
		Type:                                  datafactory.TypeAzureSQLDW,
	}
```

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	dataFactoryName := id.Path["factories"]
+	name := id.Path["linkedservices"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:210]: COuld we use a parseId function here? you can see how to generate these in resourceids.go

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	dataFactoryName := id.Path["factories"]
+	name := id.Path["linkedservices"]
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:279]: and then use the new parse function here?

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource_test.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/acceptance"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMDataFactoryLinkedServiceSynapse_ConnectionString(t *testing.T) {
```
Comment (katbyte) [inline azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource_test.go:15]: could we remove AzureRM from the test names?
```suggestion
func TestAccDataFactoryLinkedServiceSynapse_ConnectionString(t *testing.T) {
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the pr @allantargino, overall this looks good but i have left some comments on schema and resource IDs inline to address. Once they have been fixed up this should be good to merge

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go (hunk 8 above)
Comment (allantargino) [reply azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:210]: Great suggestion! I didn`t know about parse generated structures :) Thanks!

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go (hunk 1 above)
Comment (allantargino) [reply azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:64]: I like the suggestion, sounds better.

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go (hunk 3 above)
Comment (allantargino) [reply azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:76]: Also agree with this one.

File: azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go (hunk 2 above)
Comment (allantargino) [reply azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:70]: @katbyte For this one not sure if changing to `key_vault_name` represents this property well.
This property means the name of a [Key Vault linked service](https://docs.microsoft.com/en-us/azure/data-factory/store-credentials-in-key-vault#azure-key-vault-linked-service), which name can be different than your Key Vault name.
We could change from `linked_service_name` to `key_vault_linked_service` or `key_vault_linked_service_name`.
Thoughts?

Comment (katbyte) [reply azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:70]: Ah! looking more closely at the test code that makes sense, i clearly was reviewing this to quickly 😅  - agreed, it's probably fine as is, or maybe to match the resource how about `linked_service_key_vault_name` ? @allantargino

Comment (allantargino) [reply azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:70]: haha `linked_service_key_vault_name` sounds great 👍

Comment (allantargino) [reply azurerm/internal/services/datafactory/data_factory_linked_service_synapse_resource.go:70]: Actually, I will vote to stick to `linked_service_name`, as all datasets are already using this convention (example: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/data_factory_dataset_azure_blob).

**Review by katbyte — APPROVED** [review]
Thanks @allantargino - LGTM 👍

**Comment by allantargino** [comment]
Thanks for the review, @katbyte! Happy new year :)

**Pull request merged**